trieste/models/gpflow/sampler.py

```diff
@@ -630,80 +628,92 @@ def __init__(
 
         tf.debugging.assert_positive(num_features)
         self._num_features = num_features
-        self._model = model
-        feature_functions = ResampleableDecoupledFeatureFunctions(self._model, self._num_features)
+        feature_functions = ResampleableDecoupledFeatureFunctions(model, self._num_features)
 
-        super().__init__(self._model, feature_functions)
+        super().__init__(model, feature_functions)
 
-    def _prepare_weight_sampler(self) -> Callable[[int], TensorType]:
+    def _prepare_weight_sampler(self) -> Callable[[int], TensorType]:  # [B] -> [B, F + M, L]
         """
         Prepare the sampler function that provides samples of the feature weights
-        for both the RFF and cannonical feature functions, i.e. we return a function
+        for both the RFF and canonical feature functions, i.e. we return a function
         that takes in a batch size `B` and returns `B` samples for the weights of each of
-        the `L`  RFF features and `N` cannonical features.
+        the `F`  RFF features and `M` canonical features for `L` outputs.
         """
 
+        kernel = self._model.get_kernel()
+        if self._feature_functions.is_multioutput:
+
+            def kernel_K(iv: TensorType) -> TensorType:
+                # mypy is confused and doesn't see that kernel can be of two types
+                return cast(MultioutputKernel, kernel).K(iv, iv, full_output_cov=False)  # [L, M, M]
+
+        else:
+
+            def kernel_K(iv: TensorType) -> TensorType:
+                return tf.expand_dims(kernel.K(iv, iv), axis=0)  # [1, M, M]
+
         if isinstance(self._model, FeatureDecompositionInducingPointModel):
             (  # extract variational parameters
                 inducing_points,
                 q_mu,
                 q_sqrt,
                 whiten,
-            ) = self._model.get_inducing_variables()  # [M, d], [M, 1], [1, M, 1]
-            q_sqrt = q_sqrt[0, :, :]  # [M, M]
-            Kmm = self._model.get_kernel().K(inducing_points, inducing_points)  # [M, M]
+            ) = self._model.get_inducing_variables()  # [M, D], [M, L], [L, M, M], []
```

Comment:
extra []?

Reply:
Those shapes are for the 4-tuple that `get_inducing_variables()` returns. So `[]` is for `whiten` as it is a scalar.